File: SumOfSquareNumbers.java
/**
 * LeetCode 633 Sum of Square Numbers
 */
class Solution {
    public boolean judgeSquareSum(int c) {
        if (c < 0) {
            return false;
        }
        long left = 0;
        long right = (long) Math.sqrt(c);
        while (left <= right) {
            long cur = (left * left) + (right * right);
            if (cur > c) {
                right--;
            } else if (cur < c) {
                left++;
            } else {
                return true;
            }
        }
        return false;
    }
}

File: SumOfSquareNumbers.go
func judgeSquareSum(c int) bool {
	l,r := 0,int(math.Sqrt(float64(c)))
	for l <= r {
		cur := l * l + r * r
		if (cur == c) {
			return true
		} else if (cur > c) {
			r--
		} else {
			l++
		}
	}
	return false
}
